libcontainer/rootfs_linux.go

```diff
@@ -733,14 +733,20 @@ func checkProcMount(rootfs, dest string, m *mountEntry) error {
 		if m == nil {
 			return nil
 		}
-		// only allow a mount on-top of proc if it's source is "proc"
-		st, err := m.srcStatfs()
-		if err != nil {
-			return err
-		}
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// only allow a mount on-top of proc if its source is a procfs mount ...
+		if m.Device == "proc" {
```

Comment:
What I'd like is the commit message to explain why mounting procfs on top of procfs is safe. 

The commit seems confusing also, as it says it's better to be safe than sorry, but it allows more cases than before.

Reply:
Previously any mount that had a source path that was on procfs (even if it wasn't a bind-mount) was allowed because there were no checks that it was a bind-mount. This adds a check for that, and is what I'm referring to as being "better safe than sorry".

For what it's worth, the `m.Device == "proc"` is never hit anyway, so I guess we can drop it. But it is obviously safe, I don't see how being able to mount procfs on top of `/proc` is unsafe...

Reply:
I see. I wonder if we are making some mounts fail now and its worth a changelog line. IIUC a mount with a source that was a proc fs, without the "bind" option, worked before. The fix here is only requiring it to contain the option "bind".

Is this correct?

And yeah, if the m.Device == proc case can't happen, please update the code and commit message.

Reply:
Taking another look, we should be calling `checkProcMount` in more places. Adding the check to all mount types (which is correct) means that we need to check `m.Device == "proc"`.

Reply:
Uhm, this commit is changing a lot in the last few hours.

How would an example of this be look like in the config.json?

> This meant that having a non bind-mount entry for a pseudo-filesystem (like
overlayfs) with a dummy source set to /proc on the host would let you
bypass the check,

Any reason to not add a test for this?

Or even better, should this be a separate PR, with its own explanation, tests and be backported to 1.1?

Reply:
But if that is too much, just let me know when the commit/code settles :)

Reply:
There was only one new version since you last looked at it -- I added extra `checkProcMount` checks which were missing, and I moved the `m.Device == "proc"` check to the start of the function. The reasons for this change are pretty self-explanatory.

`checkProcMount` protects against mounts like 

```json
{
    "destination": "/proc",
    "type": "bind",
    "source": "/my/fake/proc",
    "options": [
        "bind"
    ]
}
```

The bug was that mounts like

```
{
    "destination": "/proc",
    "type": "overlay",
    "source": "/proc",
    "options": [
        "lowerdir=/tmp/proc1",
        "upperdir=/tmp/proc2",
        "workdir=/tmp/work"
    ]
}
```

were not blocked with the old code (note that the source path is ignored by overlayfs). 

Yes, this is a potential security issue. No, I don't think it's CVE-worthy because untrusted users shouldn't have the ability to configure arbitrary mounts (especially proper filesystem mounts, as opposed to stuff like tmpfs -- `tmpfs` is also not protected against until this commit, but `tmpcopyup` doesn't work so it's not a suitable attack method).

For what it's worth, I think we should be more strict than this. We should be checking that the inode of the mount source is equal to `PROC_ROOT_INO` (== 1). But I'm not sure if that will break real users.

There is an existing test which depends on this, in fact:

```go
func TestCheckMountDestOnProcChroot(t *testing.T) {
	m := mountEntry{
		Mount: &configs.Mount{
			Destination: "/proc",
			Source:      "foo",
			Device:      "proc",
		},
	}
	dest := "/rootfs/proc/"
	err := checkProcMount("/rootfs", dest, &m)
	if err != nil {
		t.Fatalf("destination inside proc when using chroot should not return an error: %v", err)
	}
}
```

> Any reason to not add a test for this?

I didn't think it was really that necessary, but I can add some unit tests to `libcontainer/rootfs_linux_test.go`.

> Or even better, should this be a separate PR, with its own explanation, tests and be backported to 1.1? But if that is too much, just let me know when the commit/code settles :)

I've spent several months on this PR now, I don't see a strong reason to do this. It's not as simple as a cherry-pick -- this PR changes several other aspects of `rootfs_linux.go` (including the signature of `checkProcMount` in a way that would be annoying to split and then rebase on top of) and some of them are required for this change. And I'm not sure if we will have a 1.1.11 release, given that 1.2.x is almost done once this is merged.

Reply:
SGTM, thanks